### PR TITLE
handshake-manager: matching: Generalize local matching engine logic

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR generalizes the matching engine logic for locally inferred matches. Specifically, we break out much of the core matching logic into a set of helpers in the `HandshakeExecutor` that will allow for easy re-use between the internal engine and external (atomic) engine.

This has the benefit of also simplifying and optimizing the match critical path.

I also removed the match mutual exclusion list as this argument is no longer needed.

### Testing
- All tests pass
- Live testing deferred until the feature branch is ready for merge

### CODE REVIEW

File: state/src/interface/wallet_index.rs
Comment (akirillo) [inline state/src/interface/wallet_index.rs:82]: Can we document in the doc comment that we will return a "phantom" empty balance (that doesn't actually exist in the wallet) if there is no capitalizing balance?

File: workers/handshake-manager/src/manager/matching/internal_engine.rs
Comment (akirillo) [inline workers/handshake-manager/src/manager/matching/internal_engine.rs:55]: 1. Looks like there's no `wallet.get_balance_or_default` method, so I guess you've added it in another PR, but at any rate
2. This block of code looks logically equivalent to your new `get_managed_order_and_balance` method, can use it here

File: workers/handshake-manager/src/manager/matching/internal_engine.rs
Comment (akirillo) [inline workers/handshake-manager/src/manager/matching/internal_engine.rs:105]: Have been thinking about this - do we want to shuffle? I think it's worth considering ordering either by earliest -> latest creation date, or by size, to ensure we match up the biggest fills

Comment (akirillo) [reply workers/handshake-manager/src/manager/matching/internal_engine.rs:105]: Also, there's no explicit shuffle going on here, we just insert the orders into a hash set

File: workers/handshake-manager/src/manager/matching/internal_engine.rs
Comment (akirillo) [inline workers/handshake-manager/src/manager/matching/internal_engine.rs:210]: We still need to do this check somewhere, and we no longer do. I think the best place would be in `get_match_candidates`, though it would mean two extra state lookups for wallet IDs there. conversely, it could be done in `try_settle_match`, where we already look up the wallet IDs

**Review by akirillo — APPROVED** [review]
have some comments but lgtm - just want to make sure we add back the "different wallets" check but that's a small fix so approving to unblock

File: workers/handshake-manager/src/manager/matching/internal_engine.rs
Comment (joeykraut) [reply workers/handshake-manager/src/manager/matching/internal_engine.rs:105]: yeah the shuffle is unnecessary, we can order them however we like down the line

**Pull request merged**
